File: proxy-go/handler/ratelimit.go
package handler

import (
	"math"
	"net"
	"net/http"
	"os"
	"strconv"
	"sync"
	"time"
)

// tokenBucket implements a per-IP token bucket rate limiter.
// Capacity = TSM_RATE_LIMIT (default 100) tokens, refilled at capacity/minute.
type tokenBucket struct {
	tokens   float64
	lastSeen time.Time
}

var (
	rlMu     sync.Mutex
	rlBuckets = make(map[string]*tokenBucket, 4096)
	rlCap     float64
)

func init() {
	cap := 100
	if v := os.Getenv("TSM_RATE_LIMIT"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			cap = n
		}
	}
	rlCap = float64(cap)
	// GC buckets that haven't been seen for 5 minutes.
	go func() {
		tick := time.NewTicker(5 * time.Minute)
		defer tick.Stop()
		for range tick.C {
			evictStaleBuckets()
		}
	}()
}

// checkRateLimit returns true if the request is allowed, false if limited.
func checkRateLimit(ip string) bool {
	rlMu.Lock()
	defer rlMu.Unlock()

	now := time.Now()
	b, ok := rlBuckets[ip]
	if !ok {
		rlBuckets[ip] = &tokenBucket{tokens: rlCap - 1, lastSeen: now}
		return true
	}

	elapsed := now.Sub(b.lastSeen).Minutes()
	b.lastSeen = now
	b.tokens = math.Min(rlCap, b.tokens+elapsed*rlCap)

	if b.tokens < 1 {
		return false
	}
	b.tokens--
	return true
}

func evictStaleBuckets() {
	cutoff := time.Now().Add(-5 * time.Minute)
	rlMu.Lock()
	defer rlMu.Unlock()
	for ip, b := range rlBuckets {
		if b.lastSeen.Before(cutoff) {
			delete(rlBuckets, ip)
		}
	}
}

// clientIP extracts the real IP respecting X-Forwarded-For.
func clientIP(r *http.Request) string {
	if fwd := r.Header.Get("X-Forwarded-For"); fwd != "" {
		// Take the leftmost (client) IP.
		for i := 0; i < len(fwd); i++ {
			if fwd[i] == ',' {
				return fwd[:i]
			}
		}
		return fwd
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return host
}

File: proxy/src/ratelimit.ts
/**
 * Token-bucket rate limiter — per client IP, configurable req/min.
 * Zero dependencies, in-memory, GC'd automatically on expiry.
 */

interface Bucket {
  tokens:    number;
  lastRefill: number;  // ms timestamp
}

const buckets = new Map<string, Bucket>();
const WINDOW_MS   = 60_000;                                           // 1 minute
const MAX_PER_MIN = parseInt(process.env.TSM_RATE_LIMIT ?? '100');    // configurable

/** Returns true if the request should be allowed, false if rate-limited. */
export function checkRateLimit(clientIp: string): boolean {
  const now = Date.now();
  let b = buckets.get(clientIp);

  if (!b) {
    b = { tokens: MAX_PER_MIN - 1, lastRefill: now };
    buckets.set(clientIp, b);
    return true;
  }

  // Refill tokens proportional to elapsed time
  const elapsed = now - b.lastRefill;
  const refill  = Math.floor((elapsed / WINDOW_MS) * MAX_PER_MIN);
  if (refill > 0) {
    b.tokens    = Math.min(MAX_PER_MIN, b.tokens + refill);
    b.lastRefill = now;
  }

  if (b.tokens <= 0) return false;
  b.tokens--;
  return true;
}

/** Remaining tokens for a client (for Retry-After header). */
export function remaining(clientIp: string): number {
  return buckets.get(clientIp)?.tokens ?? MAX_PER_MIN;
}

// GC stale buckets every 5 minutes
setInterval(() => {
  const cutoff = Date.now() - WINDOW_MS * 5;
  for (const [ip, b] of buckets) {
    if (b.lastRefill < cutoff) buckets.delete(ip);
  }
}, 5 * 60_000).unref();
